### DISCUSSION TITLE
A Proper Farewell, and Transparency.

### DISCUSSION THREAD

**Discussion by MotoLegacy** [nzp-team/nzportable#298] [Other Content Releases]
Hey everyone! I hope you're all doing rather well. I've noticed things have stagnated since I left, which was to be expected, but it is still unfortunate. I just want to start off with a reminder that while I won't be directly contributing to this project anymore (ever), I am still happy to be an overseer on GitHub here.

This is probably going to turn into a rant-y post, but I wanted to go into detail and share as much as I possibly could about everything being worked on, announced or not, so that nothing was left to waste. I don't want my departure to leave with it things being up in the air and content lost.

I want to establish that my departure is purely motivated in the time this project consumes, and an inability to justify it anymore. I loved working no this dearly, but after 7 years of development time, and more specifically hundreds of hours in the beginning of 2023 to get a stable PSP build ready for the development panels, and little to no fanfare, I knew I couldn't work on this anymore. This was my passion project, I love NZ:P and I love Zombies, but it's clear outside of our small little circle no one else really cared. I haven't seen anyone dump the amount of time I had, and that was discouraging, but understandable.

To start things off, here is a .7Z file with all of my unfinished mapping projects I could find. Some of these were intended to be released officially, some were just for fun (you can probably guess which were which). Lots of goodies here like "FIVE", Shi No Numa, Verruckt, etc. There might be some duplicates here, apologies if so (additionally, let me know if there's something missing here that you know I had):
[nzp_maps_archive_14052023.7z](https://archive.org/details/nzp_maps_archive_14052023.7z)

Additionally, this is a Sound Pack @DerpedCrusader had been pestering me to make and release ever since I made a video on YouTube where most sounds non-weapon were replaced with my voice. This was only a few months old. There were issues with thinking about how to release these sounds, and I grew out of the whole April Fools thing this was based on so I didn't put much work into thinking about it anyway (sorry):
[nzpsounds_myvoice.7z](https://archive.org/details/nzpsounds_myvoice.7z)

I'm not sure if I can share this, (I think I can, oh well if I cant) but this is some of the Asset source files passed over to me from blubswillrule. It just has some of the beta weapons and the M1911 which is absent from the source asset repository at the moment. Everything else is already public which is why I'm fine sharing it
 [here u go!!](https://mega.nz/file/l1sDyQZA#yqMWr9mjhvkmk6TiIaNzHxqwrEQdanG_ZDREpKL4O7g)

Okay, let's get into some of the other stuff.

Whenever I got bored Derped and I would just kinda mess around thinking about post-Black Ops content to include. One of these items were the Electric Cherry Perk-A-Cola from Black Ops II. This is why that HUD graphic is an emote in the Discord, along with others (big surprise I'm sure..). Here's a .GIF mockup I did to demonstrate how the visualization for the burst would work:

![fizz](https://github.com/nzp-team/nzportable/assets/26030250/f44dab1e-a9fb-403b-98ca-a7cac692b913)

Obviously it was going to look prettier. I didn't get far enough to think about how it'd look in the world/for co-op.

Here was the 3D model made for the machine:
[cherry.zip](https://github.com/nzp-team/nzportable/files/11474036/cherry.zip)

There's a bunch of other assets Derped made a long time ago that were being discussed for fun too (I was obsessed with the Zombie Shield and Who's Who), they've been accessible on the Discord for a while but here's the link to his Google Drive again:
[said archive](https://drive.google.com/drive/u/0/folders/0B5Se9GVjEdWYSzE5NFZyWEJoN0U)

Do NOT take that archive as a definitive list of non-Black Ops stuff. This was all stuff from Requiem, a mod I was working on in 2016/2017. It's not at all production quality and 75% of this is definitely not what I considered NZ:P-qualified feature wise. NZ:P is, at its heart, a World at War recreation. The Black Ops features that were added were all designed to be complimentary, and non-obtrusive. The exception being Kino der Toten, which I had never really liked the idea of being bundled, but I knew how much people would appreciate that map. When I think about it, including Electric Cherry, Zombie Shield, etc. is definitely a bad idea. So no one see this posting as a statement that I think it would be appropriate. :)

That's all I have to really directly show so everything else is going to be a lot of plan dumps and proper explanations as to what was going on with the project inside my head.

Let's talk about the big thing: all of the data-driving I wanted to do. Custom weapons, better Mystery Box weapon pooling, the .NZD stuff, all of it. That's still around in the branches of the [old repository](https://github.com/nzp-team/NZP-Reboot/branches). The gist with these is that on Server (and Client in the case of CSQC) start, the contents of all of these files were parsed and stored in memory somewhere. Everything I could think of that I had seen in the Call of Duty custom maps community, and in official Call of Duty maps, that could easily be added there, I made an effort to include. With weapons I had the long term plan of also including custom projectile types and defining equipment like Monkey Bombs or melee weapons. I wanted all of the stock weapons to be a part of this list to ensure there wasn't something vanilla NZ:P did that you, as a creator, could not.

I also wanted to target map compatibility with every NZ:P map, from alpha to now. That was in the `map-compatibility` branch. There isn't a whole lot to mention about that, other than the heavy lifting is re-implementing all of the vanilla Quake entities that were available in the early NZ:P days in a non-bloated fashion in order to support some of the more complicated maps. There was also a bug there with pathfinding where Zombies wouldn't find their first waypoint. That was the one bug that stopped Anstieg from being the first 100% confirmed working level. 

I wanted to revamp Perk-A-Colas in the QuakeC to make them more easily referenced outside of their designated `.QC` source files and also adding callbacks for purchasing to make adding new ones easier for modders. I'll just paste in some stuff I had written in the Discord before about that:

```
Perk functionality is stripped from server/entities/machines.qc. A new file server/entities/perks/perk_machine.qc is created. 

perk_machine.qc contains the generic functionality for a Perk. These are things like the touch function, updating the bitfield, and running the machine callbacks (explained below).

Individual perk machines are split into their own files (eg server/entities/perks/perk_revive.qc). These machine files are responsible for defining attributes specific to this machine/perk. It defines function callbacks referenced by the generic file when certain actions are committed. These callbacks are:
* give_perk_effect - used to assign values that define the perk's effect on the client
* take_perk_effect - used to reset the values that are described above
* purchase_limit_met - this is ran when (in the case of QR) the machine has been bought 3 times in solo, it would be responsible for animating the removal of the machine.
The following entity fields would be added and used to define client-affected attributes:
* revive_time_multiplier
* max_health
* barricade_rebuild_multiplier
* reload_speed_multiplier
* weapon_trace_multiplier
* weapon_firerate_multiplier
* max_stamina
* move_speed_multiplier
* explosive_damage_multiplier
* weapon_recoil_multiplier
* sniper_sway_multiplier
* aim_assist_classnames
* weapon_carrying_capacity

There are two Perk functions that are too advanced to be compatible with this idea of assigning simple client values, so that's where function overrides come in.

This is inherited from CoDs design where perks can "override" certain standard functions to execute an alternative when a certain perk is held. There would need to be two created for our perk setup:
* last_stand_override: This would be used to define Solo Revive actions instead of bleeding out in co-op
* fall_damage_override: This one would be to call upon the lack of fall damage and the explosion effect for flopper.

The perk machine entities themselves would have the following fields added to enhance customizability:
* spawnflags for the machine's power state in both solo and co-op
* bounding box size
* player viewmodel playback speed
* machine model skin
* solo purchase limit
* dynamic light level

The end goal with this design is to improve readability and modularity with our codebase, being able to find specific Perks easier, add/modify Perks without jumping around to a bunch of different places, and to allow for more community-control with how Perks act in a level.
```

Here is an old [Trello board](https://trello.com/b/AqMI9QV6/nzp-hopeful-feature-list) I had made that explains the Mystery Box changes some more, and some stuff like Shell Ejection that Derped had requested a few years ago. This hasn't been updated in some time since I had adopted [GitHub Projects](https://github.com/nzp-team/nzportable/projects/1) and used it solely for World at War parity. So there's going to be some stuff there that makes no sense.

The final thing I want to take the time to share is the Quake Protocol changes I had been making with dQuake in preparation for the PSP development panel. I had dubbed it protocol number `115`. Only dQuake and glQuake supported this at the moment, and it just made the entity modelindex (as seen [here](https://github.com/nzp-team/glquake/commit/6cc9131eb91022ccf40a9f9d777b979f008b74a8)). This needs implemented in Quakespasm and any other future non-FTE fork this project may have to keep the cross-platform multiplayer dream intact.

Oh yeah, that.

The main issue with cross-platform multiplayer was the protocol differences (Quakespasm has a bunch of fitzquake extensions and custom ones that the other ports don't). There is also an issue where the player entity struct kept having it's fields overwritten by other players. i.e., if a player went and touched a weapon trigger, the other player would get the "Hold X to buy" prompt. Never was able to diagnose that issue for sure.

I think that's all for now. Feel free to drop questions and let me know if anything is missing and I'll address it here.

xoxo,
moto/cypress

**Comment by Conmiester** [comment]
Ever since I revisited NZ:P around March-April and really enjoyed what it became nowadays than the 2011 version that I had some nostalgia of (that became lost overtime), It's a shame to see you go but thank you for everything that you worked on for NZ:P. As a CoD Zombies player, I have been really fascinated about this reboot for a long time and honestly, I hope that it would see light of the day when it becomes mostly finished + co-op support but at this point, who knows. Then again, I do would like to make some custom maps from scratch and maybe some inspired (or demake) from some custom zombie maps from World at War someday, I am aware there's a guide of it but haven't got the motivation to do it yet.

That being said, If I have any questions to say then it would be this: Will NZ;P continue on regardless of the situation? It may seem obvious at first but I'll be damned if it becomes lost in time again.

**Reply by MotoLegacy** [reply]
> Will NZ;P continue on regardless of the situation?

that's completely dependent on other people's contributions. i am happy to maintain this github organization, to manage pull requests and have discussions on people's requested changes (i've already done that a couple of times since leaving). so as long as people are still active, their changes will be brought here in an official capacity.
